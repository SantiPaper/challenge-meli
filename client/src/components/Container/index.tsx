import { FC, PropsWithChildren } from "react";
import { StyledContainer } from "./style";

export const Container: FC<PropsWithChildren> = ({ children }) => {
  return <StyledContainer>{children}</StyledContainer>;
};
