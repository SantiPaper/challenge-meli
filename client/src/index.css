@font-face {
  font-family: "Proxima nova";
  src: url("./fonts/proximanova_regular.ttf");
  font-weight: 400;
}

@font-face {
  font-family: "Proxima nova";
  src: url("./fonts/proximanova_bold.otf");
  font-weight: 600;
}

* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

html {
  font-size: 10px;
  font-family: "Proxima nova";
  background-color: var(--gray-100);
}

:root {
  --yellow: #ffe600;
  --gray-900: #333333;
  --gray-600: #666666;
  --gray-400: #999999;
  --gray-100: #eeeeee;
  --blue: #3483fa;

  --spacing-m: 1.6rem;
  --spacing-l: 3.2rem;
  --spacing-xl: 5rem;

  --font-size-100: 1.2rem;
  --font-size-200: 1.4rem;
  --font-size-300: 1.6rem;
  --font-size-400: 1.8rem;
  --font-size-500: 2.4rem;
  --font-size-600: 2.8rem;
  --font-size-700: 4.6rem;

  --radius: 0.4rem;
}

.sr-only {
  clip: rect(0 0 0 0);
  border: 0;
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}
