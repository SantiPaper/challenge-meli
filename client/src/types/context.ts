import { SearchResults } from "./items";

export type Context = {
  state: State;
  getProductsBySearch: (searchResults: SearchResults) => void;
  getSearchString: (query: string) => void;
};

export type State = {
  searchResults: SearchResults;
  searchInput: string;
};
export type Reducer = (state: State, action: Action) => State;

export type Action =
  | {
      type: "GET_PRODUCTS";
      payload: SearchResults;
    }
  | {
      type: "SEARCH_PRODUCTS";
      payload: string;
    };
