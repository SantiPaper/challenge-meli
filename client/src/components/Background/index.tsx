import { FC, PropsWithChildren } from "react";
import { Container } from "../Container";
import { StyledBackground } from "./style";

export const Background: FC<PropsWithChildren> = ({ children }) => {
  return (
    <Container>
      <StyledBackground>{children}</StyledBackground>
    </Container>
  );
};
