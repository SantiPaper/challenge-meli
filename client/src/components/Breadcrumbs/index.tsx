import { Categories } from "../../types/items";
import { StyledBreadcrumbs } from "./style";

type Props = {
  categories: Categories;
};
export const Breadcrumbs = ({ categories }: Props) => {
  return (
    <StyledBreadcrumbs>
      <ul className="breadcrumbs">
        {categories.map((catergory) => (
          <li key={catergory} className="breadcrumbs__item">
            <a href="#" className="breadcrumbs__item__link">
              {catergory}
            </a>
          </li>
        ))}
      </ul>
    </StyledBreadcrumbs>
  );
};
