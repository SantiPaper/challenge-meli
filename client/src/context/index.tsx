import { PropsWithChildren, createContext, useReducer } from "react";
import { Context } from "../types/context";
import { initialState, reducer } from "./reducer";
import { ProductDetail, SearchResults } from "../types/items";

export const ItemsContext = createContext<Context | null>(null);

export const Provider = ({ children }: PropsWithChildren) => {
  const handleFormat = (price: number) => {
    return new Intl.NumberFormat("es-AR", {
      style: "currency",
      currency: "ARS",
      minimumFractionDigits: 0,
      maximumFractionDigits: 0,
    }).format(price);
  };

  const [state, dispatch] = useReducer(reducer, initialState);

  const getProductsBySearch = (searchResults: SearchResults) => {
    dispatch({
      type: "GET_PRODUCTS",
      payload: searchResults,
    });
  };

  const getProductById = (product: ProductDetail) => {
    dispatch({
      type: "GET_PRODUCT_BY_ID",
      payload: product,
    });
  };

  return (
    <ItemsContext.Provider
      value={{
        state,
        getProductsBySearch,
        getProductById,
        handleFormat,
      }}
    >
      {children}
    </ItemsContext.Provider>
  );
};
