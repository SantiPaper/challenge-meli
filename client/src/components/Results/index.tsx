import { useSearchParams } from "react-router-dom";
import { useProductsContext } from "../../hooks/useProductsContext";
import { useFetch } from "../../hooks/useFetch";
import { Background } from "../Background";
import { ItemProduct } from "../ItemProduct";
import { Breadcrumbs } from "../Breadcrumbs";
import { Loading } from "../Loading";
import { Container } from "../Container";
import { ErrorMessage } from "../ErrorMessage";

export const Results = () => {
  const { state } = useProductsContext();
  const [params] = useSearchParams();

  const { loading, error } = useFetch({
    status: "?q=",
    payload: params.get("search"),
  });

  return (
    <section>
      {loading ? (
        <Loading />
      ) : (
        <>
          {state.searchResults.categories && (
            <Breadcrumbs categories={state.searchResults.categories} />
          )}
          {state.searchResults.items ? (
            <Background>
              {state.searchResults.items.map((prod) => {
                const number = new Intl.NumberFormat("es-AR", {
                  style: "currency",
                  currency: "ARS",
                  minimumFractionDigits: 0,
                  maximumFractionDigits: 0,
                }).format(prod.price.amount);

                return (
                  <ItemProduct
                    key={prod.id}
                    id={prod.id}
                    title={prod.title}
                    price={number}
                    shipping={prod.free_shipping}
                    img={prod.picture}
                    condition={prod.condition}
                  />
                );
              })}
            </Background>
          ) : (
            <Container>
              <ErrorMessage error={error} />
            </Container>
          )}
        </>
      )}
    </section>
  );
};
