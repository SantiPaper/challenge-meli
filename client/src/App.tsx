import { Routes, Route } from "react-router-dom";
import { Home } from "./components/Home";
import { Results } from "./components/Results";
import { Detail } from "./components/Detail";
import { Layout } from "./components/Layout";
function App() {
  return (
    <Layout>
      <Routes>
        <Route path="/" element={<Home />} />
        <Route path="/items" element={<Results />} />
        <Route path="/items/:id" element={<Detail />} />
      </Routes>
    </Layout>
  );
}

export default App;
