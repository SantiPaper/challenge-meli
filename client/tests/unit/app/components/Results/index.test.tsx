import { render, screen } from "@testing-library/react";
import { Results } from "../../../../../src/components/Results";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "../../../../../src/context";
import { server } from "../../../msw";
import { HttpResponse, http } from "msw";

describe("Componente Results", () => {
  test("Renderiza el componente", () => {
    const { container } = render(
      <Provider>
        <BrowserRouter>
          <Results />
        </BrowserRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });

  test("Renderizan los productos", async () => {
    render(
      <Provider>
        <BrowserRouter>
          <Results />
        </BrowserRouter>
      </Provider>
    );
    const cargando = screen.getByText("Cargando...");
    expect(cargando).toBeVisible();

    const firstTitle = await screen.findByRole("heading", {
      name: "Producto Genérico 1",
    });
    expect(firstTitle).toBeVisible();
    expect(cargando).not.toBeInTheDocument();

    const articles = screen.getAllByRole("article");
    expect(articles).toHaveLength(4);
  });
  test("Debería manejar un error correctamente", async () => {
    server.use(
      http.get("http://localhost:3001/api/items", () => {
        return HttpResponse.json(
          {},
          {
            status: 404,
          }
        );
      })
    );

    render(
      <Provider>
        <BrowserRouter>
          <Results />
        </BrowserRouter>
      </Provider>
    );

    const error = await screen.findByAltText("error_img");
    expect(error).toBeVisible();
  });
});
