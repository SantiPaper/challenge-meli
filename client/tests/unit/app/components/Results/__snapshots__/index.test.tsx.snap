// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Componente Results > Renderiza el componente 1`] = `
<div>
  <section
    class="sc-gLLuof dOuqnz"
  >
    <section
      class="sc-eDLKkx iQSGzz"
    >
      <h2
        class="text"
      >
        Cargando...
      </h2>
    </section>
  </section>
</div>
`;
